### PR TITLE
:sparkles: add `strict_merge` option and align list-limit semantics with Node.js qs v6.15

### PR DESCRIPTION
**Pull request by techouse**
This pull request introduces several improvements and bug fixes to the query string decoder, aligning its behavior more closely with Node.js `qs` v6.15, especially around list and object merging, duplicate handling, and array limits. The most significant changes are the addition of a new `strict_merge` option, improved handling of list limits, and enhanced duplicate key semantics. Documentation and tests have also been updated to reflect these changes.

**Decoder behavior and options:**

* Added a new `DecodeOptions.strict_merge` option (default `True`) to control how object/scalar conflicts are merged: when enabled, conflicting values are wrapped in a list (Node.js `qs` 6.15+ parity); when disabled, legacy behavior is restored, adding non-empty string scalars as object keys. [[1]](diffhunk://#diff-bcef784989b8b76801975cd7d8375e5158fa072f21b6849e6ed5747f436dd294R95-R102) [[2]](diffhunk://#diff-729a1db638a9f7b0b75f520286320950f9c1d9596ac653e86381e99c41152a9cR208-R228) [[3]](diffhunk://#diff-bd548e7db9014f8f260e91e07e0cd7656294babd0f3fecabf215b561b9e11efcR199-R230) [[4]](diffhunk://#diff-b50ea17c941d08f7cea6ca77d0a048f9215d747170852c9502457c7154c521acR16)
* Improved the handling of duplicate bracket-array assignments: bracket-array keys always combine values into arrays, regardless of the global duplicate strategy. [[1]](diffhunk://#diff-799c474eada8f1209c56fda969560ff6762067fbccf00a6f4cd500cf90a621e3L393-R420) [[2]](diffhunk://#diff-bd548e7db9014f8f260e91e07e0cd7656294babd0f3fecabf215b561b9e11efcR199-R230)
* Updated the semantics of the `list_limit` option to act as a maximum element count (matching Node.js `qs` `arrayLimit`), not a maximum index. Now, only indices below the limit are included in lists; higher indices are converted to dicts. [[1]](diffhunk://#diff-799c474eada8f1209c56fda969560ff6762067fbccf00a6f4cd500cf90a621e3L521-R547) F167923eL310R342, [[2]](diffhunk://#diff-849ca855cc7b629089e14da750ce965730c302c4a369a32077a17f5622797100L409-R413) [[3]](diffhunk://#diff-849ca855cc7b629089e14da750ce965730c302c4a369a32077a17f5622797100L715-R717) [[4]](diffhunk://#diff-24ad9527782319dbf681c523d2308cbff92ab877fd3741ccf04655c9526ecd15R105-R112)
* Improved error handling for exceeding list limits, including clearer error messages and new `CommaOverflowDict` to mark over-limit comma-split values. [[1]](diffhunk://#diff-799c474eada8f1209c56fda969560ff6762067fbccf00a6f4cd500cf90a621e3L246-R265) [[2]](diffhunk://#diff-69aa0aba847bddf11b03f8629e61538a36bf1c2c8d079ece2d3eda71b7f99e6bL13-R29) [[3]](diffhunk://#diff-729a1db638a9f7b0b75f520286320950f9c1d9596ac653e86381e99c41152a9cL526-R551) [[4]](diffhunk://#diff-729a1db638a9f7b0b75f520286320950f9c1d9596ac653e86381e99c41152a9cR579-R580)

**Documentation and tests:**

* Updated documentation in `README.rst` to describe the new `strict_merge` option, bracket-array duplicate behavior, and revised list limit semantics. [[1]](diffhunk://#diff-bd548e7db9014f8f260e91e07e0cd7656294babd0f3fecabf215b561b9e11efcR199-R230) [[2]](diffhunk://#diff-bd548e7db9014f8f260e91e07e0cd7656294babd0f3fecabf215b561b9e11efcL313-R349)
* Added and revised test cases to cover new and legacy behaviors, including strict/legacy merge, bracket-array handling, and list limit edge cases. [[1]](diffhunk://#diff-24ad9527782319dbf681c523d2308cbff92ab877fd3741ccf04655c9526ecd15R63-R73) [[2]](diffhunk://#diff-849ca855cc7b629089e14da750ce965730c302c4a369a32077a17f5622797100L409-R413) [[3]](diffhunk://#diff-849ca855cc7b629089e14da750ce965730c302c4a369a32077a17f5622797100L715-R717) [[4]](diffhunk://#diff-849ca855cc7b629089e14da750ce965730c302c4a369a32077a17f5622797100R874-R891) [[5]](diffhunk://#diff-b50ea17c941d08f7cea6ca77d0a048f9215d747170852c9502457c7154c521acR16)

**Changelog:**

* Updated `CHANGELOG.md` to document the new features and fixes.

## Summary by CodeRabbit

* **New Features**
  * Added DecodeOptions.strict_merge (defaults to True) to control object↔scalar merge outcomes.

* **Bug Fixes**
  * list_limit is a maximum element count; boundary behavior tightened.
  * Bracket-array entries always combine; over-limit comma-split results convert to an overflow object.
  * Encoding/decoding delimiter handling and dotted-key normalization aligned.

* **Documentation**
  * Clarified arrays, list limits, comma-splitting, and merge-conflict semantics.

* **Tests**
  * Expanded coverage for strict_merge, list limits, duplicates, overflow, and related behaviors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/techouse/qs_codec/pull/53?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/qs_codec/decode.py [hunk 1]
```diff
@@ -243,16 +255,14 @@ def _parse_array_value(value: t.Any, options: DecodeOptions, current_list_length
...
+            if options.raise_on_limit_exceeded:
+                raise ValueError(_list_limit_exceeded_message(options.list_limit))
+            return CommaOverflowDict({str(i): item for i, item in enumerate(split_val)})
         return split_val
 
     if options.raise_on_limit_exceeded and current_list_length >= options.list_limit:
-        raise ValueError(
-            f"List limit exceeded: Only {options.list_limit} element{'' if options.list_limit == 1 else 's'} allowed in a list."
-        )
+        raise ValueError(_list_limit_exceeded_message(options.list_limit))
```
[the remark this answers is not in the document]
Comment (techouse) [reply src/qs_codec/decode.py:263]: I checked it against Node `qs`. Copilot is right about the mechanism, but wrong if our goal is Node qs parity.

Upstream `qs` has the same check in `parseArrayValue` at parse.js (line 39):
- `qs.parse("a=b", { arrayLimit: 0, throwOnLimitExceeded: true })` raises.
- `qs.parse("a=b", { arrayLimit: -1, throwOnLimitExceeded: true })` raises.
- `qs.parse("a=", { arrayLimit: 0, throwOnLimitExceeded: true })` raises.
- `qs.parse("a", { arrayLimit: 0, throwOnLimitExceeded: true })` does not raise.

The Python behavior matches those cases with `list_limit` / `raise_on_limit_exceeded`.

So I would not change this unless we intentionally want to diverge from Node `qs`. The current behavior is odd, but it is upstream-compatible.

File: src/qs_codec/decode.py [hunk 2]
```diff
@@ -354,9 +364,11 @@ def _parse_query_string_values(
             continue
         bracket_equals_pos: int = part.find("]=")
         pos: int = part.find("=") if bracket_equals_pos == -1 else (bracket_equals_pos + 1)
+        bracket_array_assignment = pos != -1 and "[]=" in part
```
[the remark this answers is not in the document]
Comment (techouse) [reply src/qs_codec/decode.py:362]: I checked this against Node `qs`. I would not apply the suggested change if our goal is strict Node `qs` parity.

Upstream `qs` does exactly what our code does here:

- It calculates pos using `part.indexOf(']=')`, so `a=b[]=` is interpreted as key `a=b[]` and empty value.
- It then checks `part.indexOf('[]=') > -1` against the whole part, not only the key.
- That means `qs.parse("a=b[]=")` returns `{"a=b":[""]}`, and our Python code returns the same.

I also checked duplicate/comma variants:

- `a=b[]=&a=c` with duplicates: `last`/`first`/`combine`
- `a=b[]=,c` with `comma: true`
- `a=b[]=,c&a=d` with `comma: true`, `duplicates: last`

Our results match Node qs.

So the comment is reasonable from a “cleaner parser design” perspective, but it would introduce a parity divergence.

**Comment by techouse** [comment]
@coderabbitai review
